### ISSUE TITLE
Wrong initialization of `fixed_address_empty_string` causes `ShutdownProtobufLibrary` to fail

### ISSUE THREAD

**Issue by danieljoos** [protocolbuffers/protobuf#8129]
**What version of protobuf and what language are you using?**
Version: v3.14.0, master
Language: C++

**What operating system (Linux, Windows, ...) and version?**
Windows

**What runtime / compiler are you using (e.g., python version or gcc version)**
MSVC 14.27.29110
Platform Toolset v142 (Visual Studio 2019)
Debug Runtime

**What did you do?**
```
#include "addressbook.pb.h"

int main() {
	::google::protobuf::ShutdownProtobufLibrary();
}
```

with `addressbook.proto` (it really doesn't matter at all):
```
syntax = "proto3";

package tutorial;

message AddressBook {
  string name = 1;
}
```

**What did you expect to see**

No exception when execution `ShutdownProtobufLibrary()`. The program should just run-through without any issues.

**What did you see instead?**

The program throws an exception when run in Debug mode on executing `ShutdownProtobufLibrary()`.

It is caused by attempting to destruct a broken `std::string` stored `::google::protobuf::internal::fixed_address_empty_string.value`.
The reason is that the implementation relies on a static initialization of that variable. The variable `fixed_address_empty_string` is default constructed to `dummy` (here https://github.com/protocolbuffers/protobuf/blob/bd9a7104e11740e4bcfbde46c190c2d908ef331a/src/google/protobuf/message_lite.h#L158).
However, as the initialization of static global variables is undefined across translation units in C++, it is possible that `InitProtobufDefaultsImpl()` (https://github.com/protocolbuffers/protobuf/blob/bd9a7104e11740e4bcfbde46c190c2d908ef331a/src/google/protobuf/generated_message_util.cc#L82) is called before the initialization of the `fixed_address_empty_string` (https://github.com/protocolbuffers/protobuf/blob/bd9a7104e11740e4bcfbde46c190c2d908ef331a/src/google/protobuf/generated_message_util.cc#L77). This causes the allocated `std::string value` to get broken, as the `dummy` value gets initialized "over" the existing `std::string` object.

This happens, for example when including a generated protobuf header file as a call to `AddDescriptors()` is forced there to happen during dynamic initialization time.

The actual exception is thrown by the MSCV STL on destruction as the `std::string` is now malformed.

**Anything else we should know about your project / environment**

**Comment by great944** [comment]
Hello, I see this issue with windows VS2019 16.8.3.61104. is it possible to merge this fix to master please? just updated to latest release and started seeing this issue when shutting down the program with a debug build.

**Comment by danieljoos** [comment]
Hi. Unfortunately, none of the above linked commit actually fixes the issue. I was playing around in my fork and it looks like Github already associated these commits here. Not sure how it can be unlinked.

The actual issue cannot be fixed easily, imo. The problem is the `EmptyString` structure that relies on `constinit`, which is not available on MSVC.

The only viable workaround at the moment is to comment-out the `DestroyString` call for the `fixed_address_empty_string` in case of msvc debug builds.

**Comment by great944** [comment]
thank you for the update

**Comment by danieljoos** [comment]
Seems to be fixed with `v3.15.1`
